# Exploring eBay Car Sales Data

The aim of this project is to clean the data and analyze the included used car listings

importing modules

In [1]:
import pandas as pd
import numpy as np

read the file to the dataframe

In [2]:
autos = pd.read_csv("autos.csv", encoding="Latin-1")  # We need to set 
                                                #  the encoding to the Latin-1

In [3]:
autos.head()

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   
3     150000                    6   benzin  volkswagen              nein   
4      90000                    7   diesel       skoda              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

Using DataFrame.info() and DataFrame.head() methods to print information about the autos dataframe

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

What we can see:

* Dataframe contains 20 columns, most of them are string type.

* Some columns have null values, but none have more than 20% null values

* Columns use [camelcase](https://en.wikipedia.org/wiki/Camel_case) but we prefer the [snakecase](https://en.wikipedia.org/wiki/Snake_case)

We re gonna use dataframe method rename to make following changes in the dataframe:

* yearOfRegistration to registration_year

* monthOfRegistration to registration_month

* notRepairedDamage to unrepaired_damage

* dateCreated to ad_created

And in the last point to change all the names from any case to lower case we re gonna make new_columns list and replace previous columns with lower case columns

* The rest of the column names from camelcase to snakecase.

## Clean columns

In [5]:
autos.columns = autos.columns.str.lower()
autos = autos.rename(columns={"yearofregistration": "registration_year",
                              "datecrawled": "date_crawled",
                             "monthofregistration": "registration_month",
                             "notrepaireddamage": "unrepaired_damage",
                              "datacreated": "ad_created",
                              "offertype": "offer_type",
                              "vehicletype": "vehicle_type",
                              "fueltype": "fuel_type",
                              "powerps": "power_ps",
                              "datecreated": "date_created",
                              "nrofpictures": "nr_of_pictures",
                              "postalcode": "postal_code",
                              "lastseen": "last_seen",
                              "gearbox": "gear_box"
                             })

Here's our firt 5 values of dataframe

In [6]:
autos.head()

date_crawled                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat    Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

   price abtest vehicle_type  registration_year   gear_box  power_ps  model  \
0    480   test          NaN               1993    manuell         0   golf   
1  18300   test        coupe               2011    manuell       190    NaN   
2   9800   test          suv               2004  automatik       163  grand   
3   1500   test   kleinwagen               2001    manuell        75   golf   
4   3600   test   kleinwagen               2008    manuell        69  fabia   

   kilometer  registration_month fuel_type       brand unrepaired_damage  \
0     150000                   0    benzin  volkswagen               NaN   
1     125000                   5    diesel        audi                ja   
2     125000                   8    diesel        jeep               NaN   
3     150000                   6    benzin  volkswagen              nein   
4      90000                   7    diesel       skoda              nein   

          date_created  nr_of_pictures  postal_code            last_seen  
0  2016-03-24 00:00:00               0        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00               0        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00               0        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00               0        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00               0        60437  2016-04-06 10:17:21

## Cleaning the data

In [7]:
autos.describe(include = "all")

date_crawled         name  seller offer_type         price  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

        abtest vehicle_type  registration_year gear_box       power_ps  \
count   371528       333659      371528.000000   351319  371528.000000   
unique       2            8                NaN        2            NaN   
top       test    limousine                NaN  manuell            NaN   
freq    192585        95894                NaN   274214            NaN   
mean       NaN          NaN        2004.577997      NaN     115.549477   
std        NaN          NaN          92.866598      NaN     192.139578   
min        NaN          NaN        1000.000000      NaN       0.000000   
25%        NaN          NaN        1999.000000      NaN      70.000000   
50%        NaN          NaN        2003.000000      NaN     105.000000   
75%        NaN          NaN        2008.000000      NaN     150.000000   
max        NaN          NaN        9999.000000      NaN   20000.000000   

         model      kilometer  registration_month fuel_type       brand  \
count   351044  371528.000000       371528.000000    338142      371528   
unique     251            NaN                 NaN         7          40   
top       golf            NaN                 NaN    benzin  volkswagen   
freq     30070            NaN                 NaN    223857       79640   
mean       NaN  125618.688228            5.734445       NaN         NaN   
std        NaN   40112.337051            3.712412       NaN         NaN   
min        NaN    5000.000000            0.000000       NaN         NaN   
25%        NaN  125000.000000            3.000000       NaN         NaN   
50%        NaN  150000.000000            6.000000       NaN         NaN   
75%        NaN  150000.000000            9.000000       NaN         NaN   
max        NaN  150000.000000           12.000000       NaN         NaN   

       unrepaired_damage         date_created  nr_of_pictures   postal_code  \
count             299468               371528        371528.0  371528.00000   
unique                 2                  114             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq              263182                14450             NaN           NaN   
mean                 NaN                  NaN             0.0   50820.66764   
std                  NaN                  NaN             0.0   25799.08247   
min                  NaN                  NaN             0.0    1067.00000   
25%                  NaN                  NaN             0.0   30459.00000   
50%                  NaN                  NaN             0.0   49610.00000   
75%                  NaN                  NaN             0.0   71546.00000   
max                  NaN                  NaN             0.0   99998.00000   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-06 13:45:54  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                  

We can see that in columns:

* seller 

* offer type

* abtest

Almost every value is the same
nr_of_pictures columns looks odd, we need to investigate it

In [10]:
autos.nr_of_pictures.unique()

array([0], dtype=int64)

We re gonna drop this columns

In [12]:
autos.drop(columns = ["offer_type", "seller", "abtest", "nr_of_pictures"], inplace=True)

KeyError: "['offer_type' 'seller' 'abtest' 'nr_of_pictures'] not found in axis"

In [13]:
autos.head()

date_crawled                            name  price vehicle_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6    480          NaN   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  18300        coupe   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"   9800          suv   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER   1500   kleinwagen   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic   3600   kleinwagen   

   registration_year   gear_box  power_ps  model  kilometer  \
0               1993    manuell         0   golf     150000   
1               2011    manuell       190    NaN     125000   
2               2004  automatik       163  grand     125000   
3               2001    manuell        75   golf     150000   
4               2008    manuell        69  fabia      90000   

   registration_month fuel_type       brand unrepaired_damage  \
0                   0    benzin  volkswagen               NaN   
1                   5    diesel        audi                ja   
2                   8    diesel        jeep               NaN   
3                   6    benzin  volkswagen              nein   
4                   7    diesel       skoda              nein   

          date_created  postal_code            last_seen  
0  2016-03-24 00:00:00        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00        60437  2016-04-06 10:17:21

We re gonna also inspect the null values in this data frame

In [14]:
autos.isnull().sum()

date_crawled              0
name                      0
price                     0
vehicle_type          37869
registration_year         0
gear_box              20209
power_ps                  0
model                 20484
kilometer                 0
registration_month        0
fuel_type             33386
brand                     0
unrepaired_damage     72060
date_created              0
postal_code               0
last_seen                 0
dtype: int64

We can see that the columns vehicle_type, gearbox, model, fuel_type and unrepaired_damage have incomplete data

The other thing that we could see is that some non numeric-values is in German, we re gonna translate it to English

In [15]:
autos.head()

date_crawled                            name  price vehicle_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6    480          NaN   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  18300        coupe   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"   9800          suv   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER   1500   kleinwagen   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic   3600   kleinwagen   

   registration_year   gear_box  power_ps  model  kilometer  \
0               1993    manuell         0   golf     150000   
1               2011    manuell       190    NaN     125000   
2               2004  automatik       163  grand     125000   
3               2001    manuell        75   golf     150000   
4               2008    manuell        69  fabia      90000   

   registration_month fuel_type       brand unrepaired_damage  \
0                   0    benzin  volkswagen               NaN   
1                   5    diesel        audi                ja   
2                   8    diesel        jeep               NaN   
3                   6    benzin  volkswagen              nein   
4                   7    diesel       skoda              nein   

          date_created  postal_code            last_seen  
0  2016-03-24 00:00:00        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00        60437  2016-04-06 10:17:21

First let's look at the vehicle type column values

In [17]:
autos.vehicle_type.unique()

array([nan, 'coupe', 'suv', 'kleinwagen', 'limousine', 'cabrio', 'bus',
       'kombi', 'andere'], dtype=object)

In [18]:
vehicle_mapping={'limousine':'limousine','kleinwagen':'small car','kombi':'station wagon','bus':'bus','cabrio':'convertible','coupe':'coupe','suv':'suv','andere':'other','NaN':'unknown'}
autos['vehicle_type']=autos['vehicle_type'].map(vehicle_mapping)
gb_mapping={'automatik':'automatic','manuell':'manual'}
autos['gear_box']=autos['gear_box'].map(gb_mapping)
autos['unrepaired_damage'].value_counts(normalize=True,dropna=False)
upd_mapping={'nein':'no','NaN':'unknown','ja':'yes'}
autos['unrepaired_damage']=autos['unrepaired_damage'].map(upd_mapping)

In [19]:
autos.head()

date_crawled                            name  price vehicle_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6    480          NaN   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  18300        coupe   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"   9800          suv   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER   1500    small car   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic   3600    small car   

   registration_year   gear_box  power_ps  model  kilometer  \
0               1993     manual         0   golf     150000   
1               2011     manual       190    NaN     125000   
2               2004  automatic       163  grand     125000   
3               2001     manual        75   golf     150000   
4               2008     manual        69  fabia      90000   

   registration_month fuel_type       brand unrepaired_damage  \
0                   0    benzin  volkswagen               NaN   
1                   5    diesel        audi               yes   
2                   8    diesel        jeep               NaN   
3                   6    benzin  volkswagen                no   
4                   7    diesel       skoda                no   

          date_created  postal_code            last_seen  
0  2016-03-24 00:00:00        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00        60437  2016-04-06 10:17:21

## Exploring the kilometer and price columns

In [20]:
price = autos["price"]
print(price.unique().shape[0])
print()
print(price.describe())
print()
print(price.value_counts(ascending=False))
print()

5597

count    3.715280e+05
mean     1.729514e+04
std      3.587954e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      2.147484e+09
Name: price, dtype: float64

0         10778
500        5670
1500       5394
1000       4649
1200       4594
          ...  
23456         1
171000        1
21830         1
13485         1
8188          1
Name: price, Length: 5597, dtype: int64



In [29]:
autos.price.sort_values(ascending=True).unique()[:10]

array([ 0,  1,  2,  3,  4,  5,  7,  8,  9, 10], dtype=int64)

In [31]:
autos.price.sort_values(ascending=True).unique()[-10:-1]

array([10010011, 11111111, 12345678, 14000500, 27322222, 32545461,
       74185296, 99000000, 99999999], dtype=int64)

There's number of listing with prices below /30$, but there's also very large prices like ninety-nine million nine hundred ninety-nine thousand nine hundred ninety-nine. We re gonna give the broads to the car prices from 1 to half million.

In [32]:
autos = autos[autos.price.between(1,500000)]

In [33]:
autos.price.describe()

count    360650.000000
mean       5916.518938
std        9293.321687
min           1.000000
25%        1250.000000
50%        3000.000000
75%        7490.000000
max      500000.000000
Name: price, dtype: float64

In [21]:
autos.head()

date_crawled                            name  price vehicle_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6    480          NaN   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  18300        coupe   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"   9800          suv   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER   1500    small car   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic   3600    small car   

   registration_year   gear_box  power_ps  model  kilometer  \
0               1993     manual         0   golf     150000   
1               2011     manual       190    NaN     125000   
2               2004  automatic       163  grand     125000   
3               2001     manual        75   golf     150000   
4               2008     manual        69  fabia      90000   

   registration_month fuel_type       brand unrepaired_damage  \
0                   0    benzin  volkswagen               NaN   
1                   5    diesel        audi               yes   
2                   8    diesel        jeep               NaN   
3                   6    benzin  volkswagen                no   
4                   7    diesel       skoda                no   

          date_created  postal_code            last_seen  
0  2016-03-24 00:00:00        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00        60437  2016-04-06 10:17:21

In [17]:
kilometer = autos["kilometer"]
print(kilometer.unique())
print()
print(kilometer.value_counts(ascending=False))

[150000 125000  90000  40000  30000  70000   5000 100000  60000  20000
  80000  50000  10000]

150000    240797
125000     38067
100000     15920
90000      12523
80000      11053
70000       9773
60000       8669
50000       7615
5000        7069
40000       6376
30000       6041
20000       5676
10000       1949
Name: kilometer, dtype: int64


## exploring the date columns

In [18]:
autos.head()

date_crawled                            name  price vehicle_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6    480          NaN   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  18300        coupe   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"   9800          suv   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER   1500    small car   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic   3600    small car   

   registration_year   gear_box  power_ps  model  kilometer  \
0               1993     manual         0   golf     150000   
1               2011     manual       190    NaN     125000   
2               2004  automatic       163  grand     125000   
3               2001     manual        75   golf     150000   
4               2008     manual        69  fabia      90000   

   registration_month fuel_type       brand unrepaired_damage  \
0                   0    benzin  volkswagen               NaN   
1                   5    diesel        audi               yes   
2                   8    diesel        jeep               NaN   
3                   6    benzin  volkswagen                no   
4                   7    diesel       skoda                no   

          date_created  postal_code            last_seen  
0  2016-03-24 00:00:00        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00        60437  2016-04-06 10:17:21

The data set contatins 5 data columns:

* date_crawled

* registration_year

* registration_month

* date_created

* last_seen

In [24]:
autos[["date_crawled", "registration_year", "registration_month", "date_created","last_seen"]].dtypes

date_crawled          object
registration_year      int64
registration_month     int64
date_created          object
last_seen             object
dtype: object

In [34]:
autos["date_crawled"].str[:10].value_counts(normalize=True,dropna=False).sort_index(ascending=True)

2016-03-05    0.025546
2016-03-06    0.014482
2016-03-07    0.035655
2016-03-08    0.033467
2016-03-09    0.034113
2016-03-10    0.032647
2016-03-11    0.032771
2016-03-12    0.036240
2016-03-13    0.015783
2016-03-14    0.036329
2016-03-15    0.033423
2016-03-16    0.030204
2016-03-17    0.031648
2016-03-18    0.013118
2016-03-19    0.035270
2016-03-20    0.036398
2016-03-21    0.035688
2016-03-22    0.032494
2016-03-23    0.032006
2016-03-24    0.029915
2016-03-25    0.032799
2016-03-26    0.031973
2016-03-27    0.030226
2016-03-28    0.035062
2016-03-29    0.034124
2016-03-30    0.033534
2016-03-31    0.031873
2016-04-01    0.034149
2016-04-02    0.035095
2016-04-03    0.038810
2016-04-04    0.037627
2016-04-05    0.012785
2016-04-06    0.003128
2016-04-07    0.001617
Name: date_crawled, dtype: float64

The data was compiled between March and April 2016 and it's uniform

In [40]:
autos['date_created'].str[:10].value_counts(normalize=True,dropna=False).sort_index(ascending = True)

2014-03-10    0.000003
2015-03-20    0.000003
2015-06-11    0.000003
2015-06-18    0.000003
2015-08-07    0.000003
                ...   
2016-04-03    0.038999
2016-04-04    0.037735
2016-04-05    0.011618
2016-04-06    0.003119
2016-04-07    0.001553
Name: date_created, Length: 114, dtype: float64

This column is much larger and widespread in the time than the previous columns. It seems that all ads were created on website between March 2014 and April 2016

In [35]:
autos.last_seen.str[:10].value_counts(normalize=True,dropna=False).sort_index(ascending=True)

2016-03-05    0.001292
2016-03-06    0.004134
2016-03-07    0.005262
2016-03-08    0.008056
2016-03-09    0.009994
2016-03-10    0.011563
2016-03-11    0.013046
2016-03-12    0.023401
2016-03-13    0.008489
2016-03-14    0.012301
2016-03-15    0.016411
2016-03-16    0.016419
2016-03-17    0.028760
2016-03-18    0.006931
2016-03-19    0.016314
2016-03-20    0.019915
2016-03-21    0.020136
2016-03-22    0.020607
2016-03-23    0.018149
2016-03-24    0.019237
2016-03-25    0.019097
2016-03-26    0.016160
2016-03-27    0.016909
2016-03-28    0.022273
2016-03-29    0.023312
2016-03-30    0.023856
2016-03-31    0.024238
2016-04-01    0.024022
2016-04-02    0.025016
2016-04-03    0.025366
2016-04-04    0.025654
2016-04-05    0.126206
2016-04-06    0.217830
2016-04-07    0.129648
Name: last_seen, dtype: float64

It seems that the crawler was fetching data from the site between March and April 2016. This has been confirmed by both columns date_crawled, and last_seen. The distribution of this column is roughly uniform.


The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.

In [43]:
autos["registration_month"].describe()

count    360650.000000
mean          5.796046
std           3.684407
min           0.000000
25%           3.000000
50%           6.000000
75%           9.000000
max          12.000000
Name: registration_month, dtype: float64

The min value for month is 0 which suggests that the data might be missing for that particular car. The max month is 12 and it seems that the data is spread around the entire year.

In [44]:
autos.registration_year.describe()

count    360650.000000
mean       2004.433035
std          81.015348
min        1000.000000
25%        1999.000000
50%        2004.000000
75%        2008.000000
max        9999.000000
Name: registration_year, dtype: float64

The registration_year column tells about the age of the car, but according to the analysis above it contains odd values.The minimum year is 1000, when cars never existed, and the maximum year is 9999, many years into the future. This suggests that we may have false data entries and the data needs to be fixed for analysis

## Dealing with incorrect registration year

In [57]:
print(f'only {(~autos["registration_year"].between(1900,2021)).sum()/autos.shape[0] * 100}% gonna be removed')

only 0.041037016497989735% gonna be removed


it is only 4 percent, we re gonna remove these rows

In [58]:
autos = autos[autos["registration_year"].between(1900,2021)]

In [80]:
top_10_registration = autos.registration_year.value_counts().sort_values(ascending=False).head(10)

In [72]:
import matplotlib.pyplot as plt
import seaborn as sns

In [82]:
top_10_registration

2000    23122
1999    22031
2005    21725
2006    20005
2001    19745
2003    19606
2004    19475
2002    18820
2007    17506
1998    17226
Name: registration_year, dtype: int64

It shows that the most amount of all cars was registered in the begging of this century

## Explroring the price by brand

In [86]:
autos.head()

date_crawled                            name  price vehicle_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6    480          NaN   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  18300        coupe   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"   9800          suv   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER   1500    small car   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic   3600    small car   

   registration_year   gear_box  power_ps  model  kilometer  \
0               1993     manual         0   golf     150000   
1               2011     manual       190    NaN     125000   
2               2004  automatic       163  grand     125000   
3               2001     manual        75   golf     150000   
4               2008     manual        69  fabia      90000   

   registration_month fuel_type       brand unrepaired_damage  \
0                   0    benzin  volkswagen               NaN   
1                   5    diesel        audi               yes   
2                   8    diesel        jeep               NaN   
3                   6    benzin  volkswagen                no   
4                   7    diesel       skoda                no   

          date_created  postal_code            last_seen  
0  2016-03-24 00:00:00        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00        60437  2016-04-06 10:17:21

In [92]:
autos.brand.value_counts(normalize=True)

volkswagen        0.213547
bmw               0.108579
opel              0.107384
mercedes_benz     0.095988
audi              0.088710
ford              0.068907
renault           0.048338
peugeot           0.029980
fiat              0.025961
seat              0.018949
skoda             0.015465
mazda             0.015367
smart             0.014352
citroen           0.014033
nissan            0.013600
toyota            0.012827
hyundai           0.009961
sonstige_autos    0.009448
mini              0.009312
volvo             0.008982
mitsubishi        0.008219
honda             0.007584
kia               0.006926
suzuki            0.006316
alfa_romeo        0.006275
porsche           0.006039
chevrolet         0.004971
chrysler          0.003828
dacia             0.002469
jeep              0.002180
daihatsu          0.002150
land_rover        0.002103
subaru            0.002080
jaguar            0.001689
daewoo            0.001478
saab              0.001440
trabant           0.001376
l

German manufacturers represent four out of the top five brands, almost 50% of the overall listings. Volkswagen is by far the most popular brand, with approximately double the cars for sale of the next two brands combined

There are lots of brands that don't have a significant percentage of listings, so we will limit our analysis to brands representing more than 5% of total listings.

In [96]:
brand_counts = autos.brand.value_counts(normalize=True)
common_brands = brand_counts[brand_counts > 0.05].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [102]:
brand_mean_prices = {}

for brand in common_brands:
    
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    brand_mean_prices[brand] = int(mean_price)
    
brand_mean_prices

{'volkswagen': 5322,
 'bmw': 8370,
 'opel': 2940,
 'mercedes_benz': 8473,
 'audi': 8964,
 'ford': 3660}

Of the top 5 brands, there is a distinc price gap:

* Audi
* Mercedes 
* BMW

are more expensive than for example ford or opel,
to compare volkswagen is in between

## Exploring the mileage

In [105]:
bmp_series = pd.Series(brand_mean_prices)
pd.DataFrame(bmp_series, columns=["mean_price"])

mean_price
volkswagen           5322
bmw                  8370
opel                 2940
mercedes_benz        8473
audi                 8964
ford                 3660

In [112]:
brand_mean_mileage = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["kilometer"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=True)

In [114]:
brand_info = pd.DataFrame(mean_mileage, columns = ["mean_mileage"])

In [115]:
brand_info

mean_mileage
bmw                  132955
mercedes_benz        130707
audi                 129694
opel                 128930
volkswagen           128662
ford                 123915

In [116]:
brand_info["mean_prices"] = mean_prices

In [117]:
brand_info

mean_mileage  mean_prices
bmw                  132955         8370
mercedes_benz        130707         8473
audi                 129694         8964
opel                 128930         2940
volkswagen           128662         5322
ford                 123915         3660

The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.